### PR TITLE
(PUP-2080) add function to puppet language 4x

### PR DESCRIPTION
**Pull request by hlindberg**
This adds the ability to write functions in the puppet language.

### CODE REVIEW

**Comment by hlindberg** [comment]
Neeed a rebase due to someone touching egrammar.ra

File: lib/puppet/pops/loader/module_loaders.rb [hunk 1]
```diff
@@ -66,13 +69,21 @@ class AbstractPathBasedModuleLoader < Puppet::Pops::Loader::BaseLoader
...
+    def initialize(parent_loader, loaders, module_name, path, loader_name, loadables)
       super parent_loader, loader_name
 
       @module_name = module_name
       @path = path
       @smart_paths = Puppet::Pops::Loader::LoaderPaths::SmartPaths.new(self)
       @loaders = loaders
+      @loadables = loadables
+      unless (loadables - LOADABLE_KINDS).empty?
+        raise ArgumentError, "given loadables are not of supported loadable kind"
```
Comment (thallgren) [inline lib/puppet/pops/loader/module_loaders.rb]: unnecessary double qouting

File: lib/puppet/pops/loader/puppet_function_instantiator.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+class Puppet::Pops::Loader::PuppetFunctionInstantiator
+  # Produces an instance of the Function class with the given typed_name, or fails with an error if the
+  # given puppet source does not produce this instance when evaluated.
+  #
+  # @param loader [Puppet::Pops::Loader::Loader] The loader the function is associated with
+  # @param typed_name [Puppet::Pops::Loader::TypedName] the type / name of the function to load
+  # @param source_ref [URI, String] a reference to the source / origin of the puppet code to evaluate
+  # @param pp_code_string [String] puppet code in a string
+  #
+  # @return [Puppet::Pops::Functions.Function] - an instantiated function with global scope closure associated with the given loader
```
Comment (thallgren) [inline lib/puppet/pops/loader/puppet_function_instantiator.rb]: Change to `[Puppet::Pops::Function::Function]`

File: lib/puppet/pops/loader/puppet_function_instantiator.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  #
+  def self.create(loader, typed_name, source_ref, pp_code_string)
+    parser = Puppet::Pops::Parser::EvaluatingParser.new()
+
+    # parse and validate
+    result = parser.parse_string(pp_code_string, source_ref)
+    # Only one function is allowed (and no other definitions)
+    case result.model.definitions.size
+    when 0
+      raise ArgumentError, "The code loaded from #{source_ref} does not define the function #{typed_name.name} - it is empty."
```
Comment (thallgren) [inline lib/puppet/pops/loader/puppet_function_instantiator.rb]: #{typed_name.name} should probably be quoted here for clarity (3 occurrences in this file)

File: spec/fixtures/unit/pops/loaders/loaders/module_no_lib/modules/modulea/metadata.json [hunk 4]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "test-modulea",
+  "author": "test",
+  "description": "",
+  "license": "",
+  "project_page": "",
+  "source": "",
+  "summary": "",
+  "version": "1.0.0",
+  "checksums": {
```
Comment (thallgren) [inline spec/fixtures/unit/pops/loaders/loaders/module_no_lib/modules/modulea/metadata.json]: Remove 'description' and 'checksums'. They are no longer used in metadata.json

File: spec/fixtures/unit/pops/loaders/loaders/single_module/modules/modulea/metadata.json [hunk 5]
```diff
@@ -16,4 +16,4 @@
     "manifests/init.pp": "e50d8b91b632fb8357f021e6e0b6e043"
   },
```
Comment (thallgren) [inline spec/fixtures/unit/pops/loaders/loaders/single_module/modules/modulea/metadata.json]: Remove 'description' and 'checksums' while you're at it.

File: spec/unit/data_providers/function_data_provider_spec.rb [hunk 6]
```diff
@@ -47,6 +47,15 @@
     expect(resources_in(catalog)).to include(*resources_created_in_fixture)
   end
 
+  it 'it gets data from module having a puppet function delivering module data' do
```
Comment (thallgren) [inline spec/unit/data_providers/function_data_provider_spec.rb]: it 'it' tautology.

File: spec/unit/pops/loaders/loaders_spec.rb [hunk 7]
```diff
@@ -49,70 +50,132 @@
     expect(loaders.private_environment_loader().to_s).to eql("(DependencyLoader 'environment' [])")
   end
 
-  it 'can load a function using a qualified or unqualified name from a module with metadata' do
-    loaders = Puppet::Pops::Loaders.new(environment_for(module_with_metadata))
-    modulea_loader = loaders.public_loader_for_module('modulea')
+  context "when loading from a module" do
```
Comment (thallgren) [inline spec/unit/pops/loaders/loaders_spec.rb]: unnecessary double qoute

File: spec/unit/pops/loaders/loaders_spec.rb [hunk 8]
```diff
@@ -49,70 +50,132 @@
...
+      moduleb_loader = loaders.private_loader_for_module('user')
+      function = moduleb_loader.load_typed(typed_name(:function, 'user::caller')).value
+      expect(function.call({})).to eql("usee::callee() was told 'passed value' + I am user::caller()")
+
+      function = moduleb_loader.load_typed(typed_name(:function, 'user::caller')).value
+      expect(function.call({})).to eql("usee::callee() was told 'passed value' + I am user::caller()")
+    end
+  end
+
+  context "when loading from a module with metadata" do
```
Comment (thallgren) [inline spec/unit/pops/loaders/loaders_spec.rb]: unnecessary double qoute

File: spec/unit/pops/loaders/loaders_spec.rb [hunk 9]
```diff
@@ -49,70 +50,132 @@
...
+      moduleb_loader = loaders.private_loader_for_module('user')
+      function = moduleb_loader.load_typed(typed_name(:function, 'user::caller')).value
+      expect(function.call({})).to eql("usee::callee() was told 'passed value' + I am user::caller()")
+    end
   end
 
-  it 'makes dependent modules visible to a module with metadata' do
-    env = environment_for(dependent_modules_with_metadata)
-    loaders = Puppet::Pops::Loaders.new(env)
+  context "when loading from a module without metadata" do
```
Comment (thallgren) [inline spec/unit/pops/loaders/loaders_spec.rb]: double qoute...

File: spec/unit/pops/loaders/loaders_spec.rb [hunk 10]
```diff
@@ -49,70 +50,132 @@
...
+      end
+    end
+
+    it 'a 4x function can be called from a puppet function' do
+      Puppet.override({ :current_environment => scope.environment, :global_scope => scope }) do
+        function = loader.load_typed(typed_name(:function, 'user::puppetcaller4')).value
+        expect(function.call(scope)).to eql("usee::callee() got 'first' - usee::callee() got 'second'")
+      end
+    end
+
```
Comment (thallgren) [inline spec/unit/pops/loaders/loaders_spec.rb]: Missing test if a 4x Ruby function can call a puppet function.

File: spec/unit/pops/loaders/loaders_spec.rb (hunk 8 above)
Comment (hlindberg) [inline spec/unit/pops/loaders/loaders_spec.rb]: I am changing these, but I don't think we should care for things like example titles. It only really matters in performance sensitive situations.

**Pull request merged**